### ISSUE TITLE
Autocomplete cargo doc paths/links to types in doc comment

### ISSUE THREAD

**Issue by chinedufn** [intellij-rust/intellij-rust#4518]
If you write something like the following in a doc comment

```rust
/// An error returned by [`RefCell::try_borrow`](struct.RefCell.html#method.try_borrow).
#[stable(feature = "try_borrow", since = "1.13.0")]
pub struct BorrowError {
    _private: (),
}
```

The `struct.RefCell.html#method.try_borrow` your documentation site will link to the `try_borrow` method when you click on the `RefCell::try_borrow` link.

Since the naming convention for pages / links within a cargo doc generated site is consistent - could intellij-rust support autocomplete for these links?

This would make it much easier to write documentation that references other types/methods within your library.

Thanks!

**Comment by chinedufn** [comment]
On top of autocomplete it'd be nice if there were linted - i.e. maybe underlining invalid paths or something like that

**Comment by Kobzol** [comment]
This would be a really useful feature indeed. @Undin we have the individual tokens that are part of a `PsiComment` from the doc comment lexer (`RsDocElementTypes`), but I don't know what API should be used to provide name resolution for the links. I looked at `PsiReferenceContributor`, but it seems that it can only provide references for whole elements and not just their sub-parts.

**Comment by Undin** [comment]
@Kobzol `PsiReferenceContributor` can provide references inside element. For example, you recently implemented path completion inside string literals. And each path segment is a separate reference although it's still a single string literal element.
But from my point of view, the main problem is parsing. We need to parse markdown in doc comments to get proper PSI. And based on this PSI, we can provide features like name resolution, completion, inspections, etc.
We discussed with @vlad20012 how to parse markdown not to implement parsing ourselves. One possible solution is to make the same approach that used in markdown plugin. There is platform (I mean IntelliJ platform) independent parser for markdown initially written for markdown plugin (we already [use](https://github.com/intellij-rust/intellij-rust/blob/master/build.gradle.kts#L261) it for documentation rendering). And markdown plugin uses it to parse markdown file text to some tokens that are [wrapped to PSI](https://github.com/JetBrains/intellij-community/blob/master/plugins/markdown/src/org/intellij/plugins/markdown/lang/parser/MarkdownParserAdapter.java#L42-L46). We can do the same things only for doc comments. Of course, we need to solve problems that comment text contains prefixes like `///` or `//!` but looks like its possible.

**Comment by Kobzol** [comment]
>  For example, you recently implemented path completion inside string literals. And each path segment is a separate reference although it's still a single string literal element.

Oh, I didn't realize that `FileReference` is doing all of this, nice.

Markdown parsing is probably the way to go ultimately, but could we resolve links in docs even without it? The lexer already separates things like `DOC_INLINE_LINK` or `DOC_CODE_SPAN` because they have custom highlighting. Maybe we could go through the comment's lexer tokens and mark the correspondings links for resolving? But it may be too hacky :)

**Comment by Kobzol** [comment]
Some initial effort to use a Markdown parser for documentation is in this PR: https://github.com/intellij-rust/intellij-rust/pull/3739.

**Comment by Kobzol** [comment]
I tried to extend `RustParserDefinition` to create a separate parser for Markdown in doc comments. It sort-of works, but I found out that doc comments are leaf elements so I cannot easily add new children (Markdown AST nodes) to them.

I also noticed that there is already `RustDocHighlightingLexer`, which pretty much already does what we need here on the lexer side, there's just no parser/PSI connection to it yet. Maybe we could build Markdown based doc comment PSI on top of it and start adding features for Markdown PSI in doc comments and then later replace the custom lexer it with a Markdown parser?
